### ISSUE TITLE
Add a `pyproject.toml` and some guidance around backwards compatibility

### ISSUE THREAD

**Issue by rgommers** [data-apis/array-api-strict#89]
A project should have a `pyproject.toml` nowadays. I noticed it didn't after investigating a regression reported in conda-forge and seeing that the PR for 2.1.2 didn't pick up the `numpy>=2.1` version constraint in https://github.com/conda-forge/array-api-strict-feedstock/issues/12. I think it would (through Grayskull) if there would be a `pyproject.toml` to analyze.

Related: `2.1.2`, as a bugfix version, should not contain a major breaking change like that. We don't need an extensive policy I think, but a few notes in the README about what can be expected for major, minor and bugfix versions would be useful.

**Comment by asmeurer** [comment]
I think grayskull would work on setup.py metadata too. 

Regarding the version number, I didn't expect this to be that breaking. But since it evidently is, we should just revert it. As I noted elsewhere, it was just a maintenance thing. We can continue to live with the extra technical debt in this repo for a little while longer.

**Comment by rgommers** [comment]
> Regarding the version number, I didn't expect this to be that breaking. But since it evidently is, we should just revert it. As I noted elsewhere, it was just a maintenance thing.

Sounds good, let's do that first. Could you perhaps tag a new version and roll it out, then the blast radius will be small/short?

> I think grayskull would work on setup.py metadata too.

Yeah, I didn't realize it was disabled on the feedstock - will check that.

**Comment by rgommers** [comment]
I like the effort and the problem that EffVer is trying to address. There's one weird choice though: having new feature releases as micro versions and handwaving to "may overbump" is weird, so I'd expect lots of discussion around "how big is the feature, and should it be minor or micro". So you're trading one problem for another - with a lot less familiarity from the user community, hence breaking expectations of what minor and micro mean.

We're almost never going to break backwards compat, so we may end up with 2.3.15 changing versioning scheme like that. Better to stay with the current one imho.

**Comment by jacobtomlinson** [comment]
> less predictable fashion than having a main branch plus a release branch to which only safe fixes get backported.

Yeah the kind of projects that do this should stick to SemVer. But it's worth noting projects with this kind of rigour are the exception, not the norm, at least in the SciPy/PyData/DataScience communities.

> EffVer may perhaps be most suitable for very low-traffic projects where any release only contains a few changes.

I think it's more for projects what have less rigour about when PRs get merged. Many projects merge PRs whenever they are ready, and don't backport bugfixes. So the question about "which versions should this be?" comes up at release time. You're speaking with experience as someone who works on projects that merge more thoughtfully into intentional feature releases and take the time to backport bugfixes. So I would agree that EffVer isn't trying to solve your problems.

**Comment by seberg** [comment]
I suspect there are a lot of projects that adopted EffVer but do reserve the patch/micro level for bug-fix releases in practice (e.g. matplotlib).
That seems still an improvement, though.  SemVer is just not correct (breakage does occur in minor releases) while EffVer is just a bit fuzzy on the edges of when features are coming in.

That little extra promise of not adding features in minor versions seems practical to me. (As a dev, I want to back-port as little as possible since in practice anything might break things. And to me it also seems somewhat nice for users.)

But small projects without back-port branches may prefer to add a few features even into the bug-fix to signal "please update for sure, no harm done".

(If it was up to me, I would probably lean towards guidance being "micro is bug-fix only" and accept that small projects that do not back-port may ignore that so that bug-fixes are picked up.  But maybe it is just two slightly different schemes...)
